### PR TITLE
Addressing some thread safety concerns.

### PR DESCRIPTION
**Pull request by pnomolos**
- Fixed the spec so it's properly multi-threaded.  Previously the threads were being joined in the loop, so they were executing in serial.
- Changed the default pool size to 20 so the threaded tests can run
- Snychronized access to running the row function (so functions aren't simultaneously created)
- Synchronized access to the row function cache
- May fix https://github.com/seamusabshere/upsert/issues/72 - given that threading leads to random errors, it may or may not be the case.

### CODE REVIEW

File: spec/threaded_spec.rb [hunk 1]
```diff
@@ -1,18 +1,21 @@
...
-            sleep 0.2
-            upsert.row({:name => 'Jerry'}, :gender => 'male')
-            upsert.row({:name => 'Jerry'}, :gender => 'neutered')
+            ActiveRecord::Base.connection_pool.with_connection do |conn|
+              sleep 0.2
+              upsert.row({:name => 'Jerry'}, :gender => 'male')
+              upsert.row({:name => 'Jerry'}, :gender => 'neutered')
+            end
           end
-          ts.each { |t| t.join }
```
Comment (seamusabshere) [inline spec/threaded_spec.rb]: wow that is a huge error!

File: lib/upsert.rb [hunk 2]
```diff
@@ -215,19 +215,26 @@ def initialize(connection, table_name, options = {})
   #   upsert.row({:name => 'Jerry'}, :breed => 'beagle')
   #   upsert.row({:name => 'Pierre'}, :breed => 'tabby')
   def row(selector, setter = {}, options = nil)
-    row_object = Row.new(selector, setter, options)
-    merge_function(row_object).execute(row_object)
-    nil
+    @row_mutex ||= Mutex.new
```
Comment (seamusabshere) [inline lib/upsert.rb]: creating a Mutex with `||=` is itself a thread safety concern!

**Comment by pnomolos** [comment]
@seamusabshere The build is now broken ;)  I'm seeing all sorts of weird stuff locally, the most common being: `insufficient data in "T" message`

**Comment by seamusabshere** [comment]
@pnomolos revisit once #79 is merged?

**Comment by pnomolos** [comment]
@seamusabshere Yeah, that sounds like a good idea.

**Comment by pnomolos** [comment]
@seamusabshere You OK if I uncomment the mutexes now? ;)

**Comment by pnomolos** [comment]
@seamusabshere I re-added the fixes and we're passing again.  I do believe we need the mutexes.

**Comment by pnomolos** [comment]
@seamusabshere Mind re-reviewing this? :)

**Comment by seamusabshere** [comment]
@pnomolos looks good!

**Pull request merged**